# Data Science Assessment

In [2]:
#Cell to import necessary Libraries
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer

## Data Exploration
In this section, we will load the dataset into a Pandas DataFrame and perform some initial data exploration to understand the dataset better. We will check the shape of the dataset, data types of columns, display the first 5 rows of the dataset, display the last 5 rows of the dataset, display 20 random rows of the dataset, and check for null/NAN values in the dataset. This will help us get the first idea of what the Data is and what data will need to be Cleaned.


In [3]:
# Define Path to Dataset
path = "Datasets/Diabetes Dataset.csv"

# Load dataset into Pandas DataFrame
df = pd.read_csv(path)

# Display the shape of the dataset
print(f"Shape of the dataset: {df.shape}")

#Display data types of columns in the Dataset
print(f"Data types of columns in the dataset: {df.dtypes}")

# Display the first 5 rows of the dataset
print(df.head())

# display last 5 rows of the dataset
print(df.tail())

# display 20 rows of data 
print(df.sample(20))

# Display null/NAN values in the dataset
print(f"\n Null/NAN values:", df.isnull().sum())

#Display rows with 0 values in the Glucose a column
print(f"\n Rows with 0 values in the Glucose column: \n", df.loc[df['Glucose'] == 0].head())

# Display the summary statistics of the dataset
print(f"\n Summary Statistics of the dataset: \n", df.describe())




Shape of the dataset: (2768, 10)
Data types of columns in the dataset: Id                            int64
Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object
   Id  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0   1            6      148             72             35        0  33.6   
1   2            1       85             66             29        0  26.6   
2   3            8      183             64              0        0  23.3   
3   4            1       89             66             23       94  28.1   
4   5            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1         

In [4]:
# count the number of each class in the Outcome column
print(f"\n Number of each class in the Outcome column: \n", df['Outcome'].value_counts())



 Number of each class in the Outcome column: 
 Outcome
0    1816
1     952
Name: count, dtype: int64


## Continued Data Exploration
From our initial Exploration, we can see that there are no NAN/ Null values in the dataset. However, some columns have 0 in which should not be possible. We will explore which ones these are in a below cell

In [5]:
#Display rows with 0 Value in the Glucose  column
print(f"\n Examples of Rows with 0 values in the Glucose column: \n", df.loc[df['Glucose'] == 0].head())
#Display rows with 0 Value in the BloodPressure  column
print(f"\n Examples of Rows with 0 values in the BloodPressure column: \n", df.loc[df['BloodPressure'] == 0].head())
#Display rows with 0 Value in the SkinThickness  column
print(f"\n Examples of Rows with 0 values in the SkinThickness column: \n", df.loc[df['SkinThickness'] == 0].head())
#Display rows with 0 Value in the Insulin  column
print(f"\n Examples of Rows with 0 values in the Insulin column: \n", df.loc[df['Insulin'] == 0].head())
#Display rows with 0 Value in the BMI  column
print(f"\n Examples Rows with 0 values in the BMI column: \n", df.loc[df['BMI'] == 0].head())
#Display rows with 0 Value in the DiabetesPedigreeFunction  column
print(f"\n Examples of Rows with 0 values in the DiabetesPedigreeFunction column: \n", df.loc[df['DiabetesPedigreeFunction'] == 0].head())
#Display rows with 0 Value in the Age  column
print(f"\n Examples of Rows with 0 values in the Age column: \n", df.loc[df['Age'] == 0].head())




 Examples of Rows with 0 values in the Glucose column: 
       Id  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
75    76            1        0             48             20        0  24.7   
182  183            1        0             74             20       23  27.7   
342  343            1        0             68             35        0  32.0   
349  350            5        0             80             32        0  41.0   
502  503            6        0             68             41        0  39.0   

     DiabetesPedigreeFunction  Age  Outcome  
75                      0.140   22        0  
182                     0.299   21        0  
342                     0.389   22        0  
349                     0.346   37        1  
502                     0.727   41        1  

 Examples of Rows with 0 values in the BloodPressure column: 
     Id  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
7    8           10      115              0       

## Data Preprocessing
From the first look at the Dataset, we can see that some of the Columns : Skin Thickness, Insulin, BMI, Blood Pressure, and Glucose have zero values which are not possible. We will replace these zero values with NaN values and then check the proportion of missing values in the dataset, which will help us to decide on the best way to handle the missing values. This way, we can make the Data more useful for purpose of Data analytics and any predictive models

In [6]:

# Define column names of relevant columns 
columns_to_clean = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
# Duplicate DataFrame to be worked on (can be used to validate this step to check values have changed)
diabetes_data_cleaned = df.copy()
# Replace 0 with NAN value using numpy
diabetes_data_cleaned[columns_to_clean] = diabetes_data_cleaned[columns_to_clean].replace(0, np.nan)

## Data PreProcessing Validation 
This section will be used to compare the original DataFrame to the Modified dataframe from above (Note: This is just to validate that the above preprocessing worked as intended)


In [7]:
# Display the first 5 rows of the original dataset
print("\n Original Dataset: \n", df.head())
# Display the first 5 rows of the cleaned dataset
print("\n Cleaned Dataset: \n", diabetes_data_cleaned.head())

# Display the last 5 rows of the original dataset
print("\n Original Dataset: \n", df.tail())
# Display the last 5 rows of the cleaned dataset
print("\n Cleaned Dataset: \n", diabetes_data_cleaned.tail())





 Original Dataset: 
    Id  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0   1            6      148             72             35        0  33.6   
1   2            1       85             66             29        0  26.6   
2   3            8      183             64              0        0  23.3   
3   4            1       89             66             23       94  28.1   
4   5            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  

 Cleaned Dataset: 
    Id  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0   1            6    148.0           72.0           35.0      NaN  33.6   
1   2            1     85.0           66.0           29.0      NaN  26.6   
2   3   

As we can see from the output of the cell above, there are now NaN values in the cleaned dataset where there were 0 values in the original dataset. This is the desired outcome of the Data Preprocessing step.


In [8]:
#Attempt to display rows with 0 values in the glucose column of the cleaned dataset to validate the cleaning (Expected output is an empty DataFrame)
print(f"\n Examples of Rows with 0 values in the Glucose column in the cleaned dataset: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['Glucose'] == 0].head())


 Examples of Rows with 0 values in the Glucose column in the cleaned dataset: 
 Empty DataFrame
Columns: [Id, Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []


## Data Preprocessing Continued 

In [9]:
# Display the proportion of missing values in the cleaned dataset as a percentage
print(f"\n Proportion of missing values in the cleaned dataset: \n", diabetes_data_cleaned.isnull().mean() * 100)
# save cleaned dataset to csv to work on 
diabetes_data_cleaned.to_csv("Datasets/Diabetes-Dataset-Cleaned.csv", index=False)




 Proportion of missing values in the cleaned dataset: 
 Id                           0.000000
Pregnancies                  0.000000
Glucose                      0.650289
BloodPressure                4.515896
SkinThickness               28.901734
Insulin                     48.049133
BMI                          1.408960
DiabetesPedigreeFunction     0.000000
Age                          0.000000
Outcome                      0.000000
dtype: float64


As we can see from the output above, the proportion of missing values in the dataset is quite high, especially in the Skin Thickness and Insulin columns. 

## Handling Missing Values
Below we will now handle the missing values from above 



### Handling Insulin Missing Values
Due to the large percentage of missing values in his column, any values imputed will not accurately represent the dataset correctly. Due to this, we should drop the column from the dataset.

In [10]:
#handle missing Insulin values 
# Display the number of missing values in the Insulin column
print(f"\n Number of missing values in the Insulin column before imputation/Deletion: \n", diabetes_data_cleaned['Insulin'].isnull().sum())

# Drop the Insulin column from the dataset
diabetes_data_cleaned.drop(columns=['Insulin'], inplace=True)






 Number of missing values in the Insulin column before imputation/Deletion: 
 1330


### Handling Skin Thickness Missing Values
Due to the large percentage of missing values in the Skin Thickness column, any values imputed will not accurately represent the dataset correctly. Due to this, we should drop the column from the dataset.

In [11]:
#handle missing Skin Thickness values
# Display the number of missing values in the Skin Thickness column
print(f"\n Number of missing values in the Skin Thickness column before imputation/Deletion: \n", diabetes_data_cleaned['SkinThickness'].isnull().sum())

# Drop the Skin Thickness column from the dataset
diabetes_data_cleaned.drop(columns=['SkinThickness'], inplace=True)



 Number of missing values in the Skin Thickness column before imputation/Deletion: 
 800


### Handling BMI Missing Values


In [12]:
# print example rows of missing BMI values
print(f"\n Example rows of missing BMI values: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['BMI'].isnull()].head())

# Display the number of missing values in the BMI column
print(f"\n Number of missing values in the BMI column before imputation: \n", diabetes_data_cleaned['BMI'].isnull().sum())

# impute the missing values in the BMI column with the mean(average) value
imputer = SimpleImputer(strategy='mean')
diabetes_data_cleaned['BMI'] = imputer.fit_transform(diabetes_data_cleaned[['BMI']])
# Display the number of missing values in the BMI column after imputation
print(f"\n Number of missing values in the BMI column after imputation: \n", diabetes_data_cleaned['BMI'].isnull().sum())
#  print some values to validate the imputation
print(f"\n Some values in the BMI column after imputation: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['Id']== 10, 'BMI'].values[0])
print(f"\n Some values in the BMI column after imputation: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['Id']== 50, 'BMI'].values[0])
# if are the same, imputing has worked



 Example rows of missing BMI values: 
       Id  Pregnancies  Glucose  BloodPressure  BMI  DiabetesPedigreeFunction  \
9     10            8    125.0           96.0  NaN                     0.232   
49    50            7    105.0            NaN  NaN                     0.305   
60    61            2     84.0            NaN  NaN                     0.304   
81    82            2     74.0            NaN  NaN                     0.102   
145  146            0    102.0           75.0  NaN                     0.572   

     Age  Outcome  
9     54        1  
49    24        0  
60    21        0  
81    22        0  
145   21        0  

 Number of missing values in the BMI column before imputation: 
 39

 Number of missing values in the BMI column after imputation: 
 0

 Some values in the BMI column after imputation: 
 32.59666544521803

 Some values in the BMI column after imputation: 
 32.59666544521803


### Handling Blood Pressure Missing Values

In [13]:
# print example rows of missing Blood Pressure values
print(f"\n Example rows of missing Blood Pressure values: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['BloodPressure'].isnull()].head())
# take note of the id of the rows with missing values, can be used later to validate to ensure missing data handling has worked

# Display the number of missing values in the Blood Pressure column
print(f"\n Number of missing values in the Blood Pressure column before imputation: \n", diabetes_data_cleaned['BloodPressure'].isnull().sum())

# impute the missing values in the Blood Pressure column with the mean(average) value
imputer = SimpleImputer(strategy='mean')
diabetes_data_cleaned['BloodPressure'] = imputer.fit_transform(diabetes_data_cleaned[['BloodPressure']])

# Display the number of missing values in the Blood Pressure column after imputation
print(f"\n Number of missing values in the Blood Pressure column after imputation: \n", diabetes_data_cleaned['BloodPressure'].isnull().sum())




 Example rows of missing Blood Pressure values: 
     Id  Pregnancies  Glucose  BloodPressure        BMI  \
7    8           10    115.0            NaN  35.300000   
15  16            7    100.0            NaN  30.000000   
49  50            7    105.0            NaN  32.596665   
60  61            2     84.0            NaN  32.596665   
78  79            0    131.0            NaN  43.200000   

    DiabetesPedigreeFunction  Age  Outcome  
7                      0.134   29        0  
15                     0.484   32        1  
49                     0.305   24        0  
60                     0.304   21        0  
78                     0.270   26        1  

 Number of missing values in the Blood Pressure column before imputation: 
 125

 Number of missing values in the Blood Pressure column after imputation: 
 0


## Handle Glucose Missing Values

In [14]:
# finally, handle missing values in glucose column
# print example rows of missing Glucose values
print(f"\n Example rows of missing Glucose values: \n", diabetes_data_cleaned.loc[diabetes_data_cleaned['Glucose'].isnull()].head())
# take note of the id of the rows with missing values, can be used later to validate to ensure missing data handling has worked

# Display the number of missing values in the Glucose column
print(f"\n Number of missing values in the Glucose column before imputation: \n", diabetes_data_cleaned['Glucose'].isnull().sum())

# impute the missing values in the Glucose column with the mean(average) value
imputer = SimpleImputer(strategy='mean')
diabetes_data_cleaned['Glucose'] = imputer.fit_transform(diabetes_data_cleaned[['Glucose']])
# Display the number of missing values in the Glucose column after imputation
print(f"\n Number of missing values in the Glucose column after imputation: \n", diabetes_data_cleaned['Glucose'].isnull().sum())




 Example rows of missing Glucose values: 
       Id  Pregnancies  Glucose  BloodPressure   BMI  DiabetesPedigreeFunction  \
75    76            1      NaN           48.0  24.7                     0.140   
182  183            1      NaN           74.0  27.7                     0.299   
342  343            1      NaN           68.0  32.0                     0.389   
349  350            5      NaN           80.0  41.0                     0.346   
502  503            6      NaN           68.0  39.0                     0.727   

     Age  Outcome  
75    22        0  
182   21        0  
342   22        0  
349   37        1  
502   41        1  

 Number of missing values in the Glucose column before imputation: 
 18

 Number of missing values in the Glucose column after imputation: 
 0


## Handle ID Data Cleaning

In [15]:
# Drop 'Id' column from the DataFrame - not needed for analysis
diabetes_data_cleaned = diabetes_data_cleaned.drop(columns=['Id'])


In [16]:
# print percentage of missing values in the cleaned dataset

print(f"\n Proportion of missing values in the cleaned dataset: \n", diabetes_data_cleaned.isnull().mean() * 100)
# if all is 0, then all missing values have been handled correctly

#print number of missing values in the cleaned dataset
print(f"\n Number of missing values in the cleaned dataset: \n", diabetes_data_cleaned.isnull().sum())

# Finally, save the cleaned dataset to a csv file
diabetes_data_cleaned.to_csv("Datasets/Diabetes-Dataset-Cleaned-2.csv", index=False)


 Proportion of missing values in the cleaned dataset: 
 Pregnancies                 0.0
Glucose                     0.0
BloodPressure               0.0
BMI                         0.0
DiabetesPedigreeFunction    0.0
Age                         0.0
Outcome                     0.0
dtype: float64

 Number of missing values in the cleaned dataset: 
 Pregnancies                 0
Glucose                     0
BloodPressure               0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


### Skewness and Kurtosis 

In [17]:
# get skewness of the dataset
print(f"\n Skewness of the dataset: \n", diabetes_data_cleaned.skew())

# get kurtosis of the dataset
print(f"\n Kurtosis of the dataset: \n", diabetes_data_cleaned.kurtosis())

# handle skewness
# get columns with skewness greater than 1 (high skewness)
skewness = diabetes_data_cleaned.skew()
high_skewness = skewness[skewness > 1].index
print(f"\n Columns with high skewness: \n", high_skewness)

# handle kurtosis
# get columns with kurtosis greater than 3 (high kurtosis)
kurtosis = diabetes_data_cleaned.kurtosis()
high_kurtosis = kurtosis[kurtosis > 3].index
print(f"\n Columns with high kurtosis: \n", high_kurtosis)









 Skewness of the dataset: 
 Pregnancies                 0.959096
Glucose                     0.518154
BloodPressure               0.192850
BMI                         0.852292
DiabetesPedigreeFunction    1.842791
Age                         1.166299
Outcome                     0.657465
dtype: float64

 Kurtosis of the dataset: 
 Pregnancies                 0.333585
Glucose                    -0.321336
BloodPressure               1.084821
BMI                         2.406312
DiabetesPedigreeFunction    5.172935
Age                         0.771859
Outcome                    -1.568874
dtype: float64

 Columns with high skewness: 
 Index(['DiabetesPedigreeFunction', 'Age'], dtype='object')

 Columns with high kurtosis: 
 Index(['DiabetesPedigreeFunction'], dtype='object')


From the Skewness, we can see that 2 Columns, DiabetesPedigreeFunction and Age have high skewness.   
From the Kurtosis, we can see that 2 columns, DiabetesPedigreeFunction and BMI have high kurtosis.(While above 3 is usually considered higher, BMI(2.4) seems high enough where it seems worth it to add into this category)